### PR TITLE
Fix artificial rules leak

### PR DESCRIPTION
**Pull request by zurk**
Fixes https://github.com/src-d/style-analyzer/issues/375
Also, fix Rule filtering. It was not applied. 
Fix strange behavior when analyzer tries to replace everything with `␣'`

Improve code a little bit.

### CODE REVIEW

File: lookout/style/format/research/seq_matching_vis.py [hunk 1]
```diff
@@ -50,7 +50,8 @@ def visualize(input_filename: str, bblfsh: str, language: str, model_path: str)
     fe = FeatureExtractor(language=language, **rules.origin_config["feature_extractor"])
     X, y, vnodes_y, vnodes = fe.extract_features([file])
 
-    y_pred, _ = rules.predict(X, vnodes_y, vnodes, fe)
+    # TODO: fix rule visualization if new rules were added
+    y_pred, _, new_rules = rules.predict(X, vnodes_y, vnodes, fe)
```
[the remark this answers is not in the document]
Comment (m09) [reply lookout/style/format/research/seq_matching_vis.py]: As far as I know this code as not been used. @EgorBu used it to experiment for a few days when we were building the evaluation datasets but dropped it if I recall correctly. It's not used in visualization at least.

File: lookout/style/format/rules.py [hunk 2]
```diff
@@ -173,48 +166,43 @@ def _get_composite(feature_extractor: FeatureExtractor, labels: Tuple[int, ...])
...
         """
         quotes_classes = {CLASS_INDEX[CLS_DOUBLE_QUOTE], CLASS_INDEX[CLS_SINGLE_QUOTE]}
+        processed_rules = list(self.rules)
         processed_y = y_pred.copy()
         processed_winners = winners.copy()
-        y_indices = {id(vnode): i for i, vnode in enumerate(vnodes_y)}
         new_rules = {}
+
+        def append_new_rule(labels: Tuple[int, ...], y_i: int, conf: float, support: int) -> None:
+            if labels in new_rules:
```
Comment (m09) [inline lookout/style/format/rules.py]: this does not assign the correct stats if a previous rule on the same labels was created with different stats.

File: lookout/style/format/rules.py [hunk 3]
```diff
@@ -173,48 +166,43 @@ def _get_composite(feature_extractor: FeatureExtractor, labels: Tuple[int, ...])
...
-        rule = self._get_new_rule(feature_extractor, tuple(labels), new_rules)
-        winners[y_i] = rule
-        y[y_i] = self._rules[rule].stats.cls
-
-    def harmonize_quotes(self, X: numpy.ndarray, y_pred: numpy.ndarray,
+    def harmonize_quotes(self, y_pred: numpy.ndarray,
                          vnodes_y: Sequence[VirtualNode], vnodes: Sequence[VirtualNode],
                          winners: numpy.ndarray, feature_extractor: FeatureExtractor
-                         ) -> Tuple[numpy.ndarray, numpy.ndarray]:
+                         ) -> Tuple[numpy.ndarray, numpy.ndarray, List[Rule]]:
```
Comment (m09) [inline lookout/style/format/rules.py]: We should return a complete `Rules` instance

File: lookout/style/format/visualizer/server.py [hunk 4]
```diff
@@ -106,7 +106,7 @@ def return_features() -> Response:
                     "ground_truths": y.tolist(),
                     "predictions": y_pred.tolist(),
                     "sibling_indices": sibling_indices,
-                    "rules": describe_rules(rules.rules, fe),
+                    "rules": describe_rules(new_rules, fe),
```
[the remark this answers is not in the document]
Comment (m09) [reply lookout/style/format/visualizer/server.py]: yup! When the function will return a `Rules` instance we can even just override rules in the rules.predict call and not touch this line :)

File: lookout/style/format/research/seq_matching_vis.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (zurk) [reply lookout/style/format/research/seq_matching_vis.py]: wow, I did not pay attention it is in research directory. 😄

File: lookout/style/format/rules.py (hunk 2 above)
Comment (zurk) [reply lookout/style/format/rules.py]: Yeah. do you think it is a good idea to create a new rule here too?

Comment (m09) [reply lookout/style/format/rules.py]: yes sadly it is necessary to have as many rules as (cls, precision, tuple) comboes :(

Comment (zurk) [reply lookout/style/format/rules.py]: So there is a case when we can reuse rule. They goes from 
```python
            if labels1[-1] not in quotes_classes or labels3[0] not in quotes_classes:
                append_new_rule(vnode1.y, y_i_1, 1., 1)
                append_new_rule(vnode3.y, y_i_3, 1., 1)
```

so it is better to keep the logic, but chage `new_rules` indexes.  that what I did

File: lookout/style/format/visualizer/server.py [hunk 5]
```diff
@@ -106,7 +106,7 @@ def return_features() -> Response:
                     "ground_truths": y.tolist(),
                     "predictions": y_pred.tolist(),
                     "sibling_indices": sibling_indices,
-                    "rules": describe_rules(rules.rules, fe),
+                    "rules": describe_rules(rules, fe),
```
Comment (m09) [inline lookout/style/format/visualizer/server.py]: this needs to be changed back to `rules.rules` now that `predict` returns a `Rules` and not `Sequence[Rule]`

Comment (m09) [reply lookout/style/format/visualizer/server.py]: just noticed Vadim also made a comment on that line, sorry for the duplicate

**Comment by zurk** [comment]
should be good now

**Pull request merged**